File: soundcloud/soundcloud.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys, traceback
import os, getopt
from httplib import HTTP
import simplejson as json
from urlparse import urlparse
import urllib, urllib2
import socket
from xml.dom import minidom
from struct import unpack, pack


def soundex(name, len=4):
    """ soundex module conforming to Knuth's algorithm
        implementation 2000-12-24 by Gregory Jorgensen
        public domain
    """
    digits = '01230120022455012623010202'
    sndx = ''
    fc = ''

    for c in name.upper():
        if c.isalpha():
            if not fc: fc = c
            d = digits[ord(c)-ord('A')]
            if not sndx or (d != sndx[-1]):
                sndx += d

    sndx = fc + sndx[1:]

    sndx = sndx.replace('0','')

    return (sndx + (len * '0'))[:len]

def print_result(o):
	s = str(json.dumps(o))
	debugMsg("%s" % s)
	sys.stdout.write(pack('>L', len(s)))
	sys.stdout.write(s)
	sys.stdout.flush()

def debugMsg(message):
	pass

class TomahawkResolver:
	def results(self, query):
		return []
	
	def resolver_settings(self):
		return {}
		
	def settings(self):
		debugMsg("settings")
		settings =  {
			'_msgtype':'settings', 
			'name':'Generic Python Resolver', 
			'targettime':10000, 
			'weight':80
		}
		settings.update(self.resolver_settings())
		debugMsg(settings)
		return settings
		
	def resolve(self, query):
		return {'_msgtype':'results', 'qid': query['qid'], 'results':self.results(query)}
	
	@classmethod
	def start_static(cls):
		return cls().start()
		
	def start(self):
		print_result(self.settings())
		
		while 1:
			length = sys.stdin.read(4)
		
			if not length:
				break;

			length = unpack('!L', length)[0]
			if not length:
				break
			if length > 4096 or length < 0:
				break
			if length > 0:
				msg = sys.stdin.read(length)
				try:
					request = json.loads(msg)
					results = self.resolve(request)
					print_result(results)
				except:
					traceback.print_exc(file=sys.stderr)
					pass

class SoundCloudResolver(TomahawkResolver):
        def resolver_settings(self):
            return {'name':"SoundCloud Resolver", 'targettime':10000, 'weight':80}

        def results(self, query):
            param_string = ''
            params = {}
            params['filter'] = "streamable"
            params['q'] = '"'+query['artist']+' '+query['track']+'"';
            param_string = urllib.urlencode(params)

            url = 'http://api.soundcloud.com/tracks.json?client_id=TiNg2DRYhBnp01DA3zNag&'+param_string

            c=urllib2.urlopen(url)
            contents = c.read()
            try:
                data = json.loads(contents)
            except ValueError:
                return []

            if len(data) < 1:
                return []
            
            i =0
            song = {};
            song['artist'] = query['artist']
            song['track'] = data[i]['title']
	    #song['artist'] = data[i]['title'].split(" - ")[0]
	    #song['track'] = data[i]['title'].split(" - ")[1]
	    song['source'] = 'SoundCloud'
            song['score'] = 1.00;
            song['url'] = data[i]['stream_url']+".json?client_id=TiNg2DRYhBnp01DA3zNag"
            song['duration']= data[i]['duration']/1000
            song['mimetype'] = 'audio/mpeg'
            song['bitrate'] = 128;
            song['year'] = data[i]['release_year']
            return [song];
        def test(self):
            print self.results({'artist':"Bon Jovi", 'track':"You Give Love A Bad Name"})
                
if __name__ == "__main__":
    try:
        opts, args = getopt.getopt(sys.argv[1:], "t", ["test"])
        for opt, arg in opts:
            if opt in ("-t", "--test"):
                r = SoundCloudResolver()
                r.test()
                sys.exit()
    except getopt.GetoptError:
        sys.exit(2)

    SoundCloudResolver.start_static()

File: soundcloud/soundcloud.js
var SoundcloudResolver = Tomahawk.extend(TomahawkResolver, {
	
	getConfigUi: function () {
		var uiData = Tomahawk.readBase64("config.ui");
		return {
			"widget": uiData,
			fields: [{
				name: "includeCovers",
				widget: "covers",
				property: "checked"
			}, {
				name: "includeRemixes",
				widget: "remixes",
				property: "checked"
			}, {
				name: "includeLive",
				widget: "live",
				property: "checked"
			}],
			images: [{
				"soundcloud.png" : Tomahawk.readBase64("soundcloud.png")
			}]
		};
	},

	newConfigSaved: function () {
		var userConfig = this.getUserConfig();
		if((userConfig.includeCovers != this.includeCovers) || (userConfig.includeRemixes != this.includeRemixes) || (userConfig.includeLive != this.includeLive)) {
			this.includeCovers = userConfig.includeCovers;
			Tomahawk.log("Include Covers is set to: " + this.includeCovers);
			this.includeRemixes = userConfig.includeRemixes;
			Tomahawk.log("Include Remixes is set to: " + this.includeRemixes);
			this.includeLive = userConfig.includeLive;
			Tomahawk.log("Include Live is set to: " + this.includeLive);
		}
	},
	
	settings: {
		name: 'Soundcloud',
		weight: 85,
		timeout: 5
	},
	
	init: function() {
		String.prototype.capitalize = function(){
		return this.replace( /(^|\s)([a-z])/g , function(m,p1,p2){ return p1+p2.toUpperCase(); } );
		};
	},	

	getTrack: function (trackTitle, origTitle) {
		if ((this.includeCovers === false || this.includeCovers === undefined) && trackTitle.search(/cover/i) !== -1 && origTitle.search(/cover/i) === -1){
			return null;
		}
		if ((this.includeRemixes === false || this.includeRemixes === undefined) && trackTitle.search(/remix/i) !== -1 && origTitle.search(/remix/i) === -1){
			return null;
		}
		if ((this.includeLive === false || this.includeLive === undefined) && trackTitle.search(/live/i) !== -1 && origTitle.search(/live/i) === -1){
			return null;
		}
		else {
			return trackTitle;
		}
	},

	searchSoundcloud: function( qid, query, title, artist ) {
		var apiQuery = "http://api.soundcloud.com/tracks.json?consumer_key=TiNg2DRYhBnp01DA3zNag&filter=streamable&q=" + query;
		var that = this;
		Tomahawk.asyncRequest(apiQuery, function(xhr) {
			var resp = JSON.parse(xhr.responseText);
			if (resp.length === 0){
				return;
			}
			
			var results = [];
			var stop = 0;
			if (title !== ""){ 
				stop = 1;
			}
			else{ 
				stop = resp.length;
			}
			for (i = 0; i < stop; i++) {
				// Need some more validation here
				// This doesnt help it seems, or it just throws the error anyhow, and skips?
				if(resp[i] === undefined){
					continue;
				}
	
				// Check whether the artist and title (if set) are in the returned title, discard otherwise
				if (resp[i].title !== undefined && resp[i].title.toLowerCase().indexOf(artist.toLowerCase()) === -1 || (title !== "" && resp[i].title.toLowerCase().indexOf(title.toLowerCase()) === -1)) {
					continue;
				}
				var result = new Object();
				if (artist !== "") {
					result.artist = artist;
				}
				if (that.getTrack(resp[i].title, title)){
					if (title !== ""){
						result.track = title;
					}
					else {
						if (resp[i].title.toLowerCase().indexOf(artist.toLowerCase() + " - ") === 0){
							result.track = resp[i].title.slice((artist.toLowerCase() + " - ").length);
						}
						else if (resp[i].title.toLowerCase().indexOf(artist.toLowerCase() + " — ") === 0){
							result.track = resp[i].title.slice((artist.toLowerCase() + " — ").length);
						}
						else {
							result.track = resp[i].title; // remove artist from here
						}
					}
				}
				else {
					continue;
				}
	
				
				result.source = that.settings.name;
				result.mimetype = "audio/mpeg";
				result.bitrate = 128;
				result.duration = resp[i].duration / 1000;
				result.score = 0.85;
				result.year = resp[i].release_year;
				result.url = resp[i].stream_url + ".json?client_id=TiNg2DRYhBnp01DA3zNag";
				results.push(result);
			}
			var return1 = {
				qid: qid,
				results: results
			};
			Tomahawk.addTrackResults(return1);
		});
	},

	resolve: function(qid, artist, album, title)
	{
		if (artist !== "") {
			query = encodeURIComponent(artist) + "+";
		}
		if (title !== "") {
			query += encodeURIComponent(title);
		}
		this.searchSoundcloud(qid, query, title, artist);
	},
	
	search: function( qid, searchString )
	{
		// First we get the artist name out of the search string with echonest's artist/extract function
		// HACK because Echo Nest is case-sensitive for artist/extract. we capitalize all words in the query just as a precaution. maybe a bad idea..
		// NOTE that this can often be slow, so the results can time out. However, Tomahawk can't deal with results without artists, so we *need an artist, so we try anyway
		searchString = encodeURIComponent(searchString.capitalize());
		var url = "http://developer.echonest.com/api/v4/artist/extract?api_key=JRIHWEP6GPOER2QQ6&format=json&results=1&sort=hotttnesss-desc&text=" + searchString;
		var that = this;
		Tomahawk.asyncRequest(url, function(xhr) {
			var response = JSON.parse(xhr.responseText).response;
			var artist = "";
			if (response && response.artists && response.artists.length > 0) {
				artist = response.artists[0].name;
				
				Tomahawk.log("Found artist " + artist + " for search string \"" + searchString + "\"");
				
				that.searchSoundcloud(qid, searchString, "", artist);
			}
		});
	}
});

Tomahawk.resolver.instance = SoundcloudResolver;